File: public/js/alarm-editor.js
$(document).ready(function () {
    updateEveryDayCheckBox();
    // autocomplete for tracks
//    $.get("/tracks", function (tracks) {
//        var trackIDs = jQuery.map(tracks, function (n, i) {
//            return {
//                label: n.title,
//                value: n.title,
//                id: n.id
//            };
//        });
    $(".selector").autocomplete({
        source: function (request, response) {
            $.getJSON("/search?f=json", request, function (data, status, xhr) {
                response($.map(data, function (v, i) {
                    return {
                        label: v.artist + " - " + v.title,
                        value: v.artist + " - " + v.title,
                        id: v.id
                    }
                }));
            });
        },
        minLength: 3,
        select: function (event, ui) {
            var item = ui.item;
            // stores the track ID in a hidden field
            $("#track_id").val(item.id);
        }
    });
//    });
});

File: frontend/src/main/scala/org/musicpimp/js/AlarmEditor.scala
package org.musicpimp.js

import org.scalajs.jquery.{JQueryEventObject, JQueryXHR, jQuery}

import scala.concurrent.{Future, Promise}
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.JSON

class AlarmEditor extends BaseScript {
  val pimpQuery = global.jQuery.asInstanceOf[PimpQuery]

  val EveryId = "every"
  val Checked = "checked"
  val CheckedSelector = ":checked"
  val everyDay = Seq("mon", "tue", "wed", "thu", "fri", "sat", "sun")
  val trackIdElem = elem("track_id")

  elem(EveryId).click((_: JQueryEventObject) => onEveryDayClicked())
  everyDay.map(elem) foreach { e =>
    e.click((_: JQueryEventObject) => updateEveryDayCheckbox())
  }
  val autoOptionsFuture = AutoOptions.fromAsync(
    req => searchFuture[Track](req).map(ts => ts.map(AutoItem.from)),
    item => trackIdElem.value(item.id))
  val ui: JQueryUI = jQuery(".selector")
  ui.autocomplete(autoOptionsFuture)
  updateEveryDayCheckbox()

  def onEveryDayClicked() = {
    val newValue = elem(EveryId).is(CheckedSelector)
    setAll(everyDay, newValue)
  }

  def updateEveryDayCheckbox() = {
    val isEveryDayClicked = everyDay.forall(isChecked)
    setAll(Seq(EveryId), isEveryDayClicked)
  }

  def searchFuture[T: PimpJSON.Reader](term: Request): Future[Seq[T]] = {
    val p = Promise[Seq[T]]()
    search[T](term)(ts => p.success(ts))
    p.future
  }

  def search[T: PimpJSON.Reader](term: Request)(onResults: Seq[T] => Unit) =
    PimpQuery.getJSON(
      "/search?f=json",
      term,
      response => {
        val results = PimpJSON.read[Seq[T]](response.body)
        onResults(results)
      })

  def isChecked(id: String) = elem(id).is(CheckedSelector)

  def setAll(ids: Seq[String], value: Boolean) =
    ids.foreach(id => elem(id).prop(Checked, value))
}
